### PR TITLE
opt: rewrite solo mode in the dictionary bar

### PR DESCRIPTION
**Pull request by xiaoyifang**
the differences after the rewrite:

For example, there are 4 dictionaries A,B,C,D with ABC selected.
| Cases| Before| After|  Note|
|--------|--------|--------|--------|
| Ctrl+Click A| select A only| select A only|
| Ctrl+Click A, Ctrl+Click B | select B only| select B only|
| Ctrl+Click A, Ctrl+Click A | A,B,C,D selected(all dictionaries selected)| all dictionaries selected| 
| Ctrl+Click A, Shift+Click A| A,B,C selected | A,B,C selected|
| Ctrl+Click A, Ctrl+Click B, Shift+Click B| A selected|  A,B, C selected|  &NotEqual;
| Shift+Click A| A selected|  no effect,A,B, C selected|  &NotEqual;
| Shift+Click A,Shift+Click A| A B C selected|  A,B, C selected| 
| Shift+Click A,Shift+Click B|B selected|  A,B, C selected| &NotEqual;

to summarize the result.
After the rewrite:
Ctrl enter the solo mode ,Shift restore .

### CODE REVIEW

**Comment by xiaoyifang** [comment]
> The Ctrl was to switch between One <-> All

this behavior was preserved.

**Comment by xiaoyifang** [comment]
> Btw, there is a bug in the new code: If previously all dicts are selected, Ctrl + Click A will select A, but then Shift click won't restore back to have all dicts selected.

fixed now
